analysis/skeleton_length_analysis.py

```diff
@@ -0,0 +1,217 @@
+"""Analyzing skeleton length (relative to demos) generated by task planning as
+a function of the number of predicates included."""
+
+from typing import Sequence, Set
+import os
+import numpy as np
+from numpy.typing import NDArray
+import matplotlib.pyplot as plt
+from predicators.src.datasets import create_dataset
+from predicators.src.envs import create_env
+from predicators.src.approaches import ApproachFailure, ApproachTimeout
+from predicators.src import utils
+from predicators.src.settings import CFG
+from predicators.src.planning import task_plan, task_plan_grounding
+from predicators.src.structs import Dataset, Task, Predicate
+from predicators.src.nsrt_learning import segment_trajectory, \
+    learn_strips_operators
+
+ENV_NAMES = ["cover", "blocks", "painting"]
+
+SEEDS = list(range(10))
+
+
+def _run_analysis_for_env(env_name: str,
+                          seed: int,
+                          max_skeletons: int = 8,
+                          timeout: float = 10,
+                          error_upper_bound: int = 100) -> NDArray[np.int32]:
+    # Create data for this environment and seed.
+    utils.reset_config({"seed": seed, "env": env_name})
+    env = create_env(env_name)
+    env.seed(seed)
+    train_tasks = env.get_train_tasks()
+    dataset = create_dataset(env, train_tasks)
+    assert [t.train_task_idx for t in dataset.trajectories] == \
+        list(range(len(train_tasks)))
+    demo_skeleton_lengths = [
+        utils.num_options_in_action_sequence(t.actions)
+        for t in dataset.trajectories
+    ]
+    # Get the oracle predicates.
+    oracle_predicates = env.predicates
+    # Starting with an empty predicate set (plus goal predicates),
+    # build up to the oracle predicate set, adding one predicate at a time.
+    current_predicate_set = set(env.goal_predicates)
+    num_candidates = len(oracle_predicates) - len(current_predicate_set)
+    # All results, final shape (num candidates + 1, num tasks, max skeletons).
+    all_results = []
+    # Start by evaluating the initial current_predicate set.
+    initial_predicate_set_result = _compute_skeleton_length_errors(
+        dataset,
+        train_tasks,
+        current_predicate_set,
+        demo_skeleton_lengths,
+        seed,
+        max_skeletons=max_skeletons,
+        timeout=timeout,
+        error_upper_bound=error_upper_bound)
+    all_results.append(initial_predicate_set_result)
+    for _ in range(num_candidates):
+        # Evaluate each possible next candidate predicate set and keep the
+        # best one.
+        best_next_predicate_set = set()
+        # This is just for type checking; it will get overwritten.
+        best_next_predicate_set_result = initial_predicate_set_result
```

Comment:
is all this info really just for making the heatmap? for the graphs you just need the scores right?

Reply:
these are just scores, they're just not averaged / min'd out yet. you're right that if we just made the graphs we could average / min immediately